### DISCUSSION TITLE
Confusion with git fetch and git pull

### DISCUSSION THREAD

**Discussion by senior-software-engineers** [community/community#135561] [Pull Requests] [answered]
### Select Topic Area

Question

### Body

I was reading the Git SCM book, specifically the section on pulling, and it states:

The `git fetch` command retrieves all the changes from the server that you haven't yet downloaded, but it does not alter your working directory. It simply gathers the data for you, allowing you to merge it at your discretion.

However, after using `git fetch upstream`, I noticed that it seems to merge all the changes from the remote branch into my local branch, which makes it appear as though my working directory has been updated. This seems to contradict the earlier statement.

I'm feeling confused and having trouble understanding this. Could someone clarify this for me?

**Update:**

Here’s the command I executed along with its output:

```bash
bash-3.2$ git fetch upstream
remote: Counting objects: 108, done.
remote: Compressing objects: 100% (8/8), done.
remote: Total 108 (delta 77), reused 77 (delta 77), pack-reused 23
Receiving objects: 100% (108/108), 25.92 KiB | 0 bytes/s, done.
Resolving deltas: 100% (79/79), completed with 31 local objects.
From <git-repo-url>
   a82339d..9844eeb  master -> upstream/master
   802bae5..6c84bfb  <some-branch> -> upstream/<some-branch>
```

**Comment by poyrazavsever** [comment]


**Reply by senior-software-engineers** [reply]
This is AI answer.

**Comment by everforyou** [comment, marked as answer]
Hi @skystarxtogether 
I used the command `git fetch upstream`, and it seems to merge all the changes from the remote branch into my local branch, which I thought also updated my working directory. However, this appears to contradict the earlier statement.

The `git fetch` command does not update your local branches or your working directory.

You might assume it updates a local branch because of a line like this:
```bash
   a82339d..9844eeb  master -> upstream/master
```
But according to the Output section in the `git fetch` manual, this indicates that:

- "master" refers to the name of the "ref" (with a branch being one type of ref) in the remote repository being fetched from, and
- "upstream/master" is the name of the ref in the local repository that is being updated.

It's important to note that the "master" branch in your local repository is not updated.

**Comment by LavishT04** [comment]
When you run git fetch upstream, Git downloads all new commits from the upstream remote that aren't in your local repository.
It updates the remote-tracking branches in your local repository (like upstream/master, upstream/<some-branch>). These are references to the state of branches on the remote.
However, it does not merge these changes into your local branches or update your working directory.

```
a82339d..9844eeb  master -> upstream/master
802bae5..6c84bfb  <some-branch> -> upstream/<some-branch>
```


This indicates that the remote-tracking branches upstream/master and upstream/<some-branch> have been updated to new commits, but your local master and <some-branch> have not been changed.
If you're seeing changes in your working directory after a fetch, it could be due to one of these reasons:

You might be confusing git fetch with git pull. git pull is essentially git fetch followed by git merge.
You might have some Git hooks or aliases set up that automatically merge after a fetch.
You might be looking at the output of git status after the fetch, which shows how your branch relates to the updated remote-tracking branch, which can sometimes be misleading.

To verify that git fetch isn't changing your working directory, you can:

Check the state of your working directory before and after the fetch using git status.
Look at your current branch's commit using git log -1 before and after the fetch.

If these remain the same after git fetch, then your working directory and current branch haven't been changed, even though the remote-tracking branches have been updated.
